pkg/volume/util/nested_volumes.go

```diff
@@ -0,0 +1,68 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package util
+
+import (
+	"fmt"
+	"k8s.io/api/core/v1"
+	"os"
+	"path"
+	"path/filepath"
+	"sort"
+	"strings"
+)
+
+// MakeNestedMountpoints creates mount points in baseDir for volumes mounted beneath name
+func MakeNestedMountpoints(name, baseDir string, pod v1.Pod) error {
+	for i := 0; i < len(pod.Spec.InitContainers)+len(pod.Spec.Containers); i++ {
+		var container *v1.Container
+		if i < len(pod.Spec.InitContainers) {
+			container = &pod.Spec.InitContainers[i]
+		} else {
+			container = &pod.Spec.Containers[i-len(pod.Spec.InitContainers)]
+		}
+		var mountPoints []string
+		var myMountPoints []string
+		for _, vol := range container.VolumeMounts {
+			cleaned := filepath.Clean(vol.MountPath)
+			mountPoints = append(mountPoints, cleaned)
+			if vol.Name == name {
+				myMountPoints = append(myMountPoints, cleaned)
+			}
+		}
+		sort.Strings(mountPoints)
+		parentPrefix := ".." + string(os.PathSeparator)
+		for _, mountPoint := range myMountPoints {
+			if strings.HasPrefix(mountPoint, parentPrefix) {
+				// Don't let a container trick us into creating directories outside of its rootfs
+				return fmt.Errorf("Invalid container mount point %v", mountPoint)
+			}
+			mpSlash := mountPoint + string(os.PathSeparator)
+			for prev, cur := -1, sort.SearchStrings(mountPoints, mpSlash); cur < len(mountPoints) && strings.HasPrefix(mountPoints[cur], mpSlash); cur++ {
```

Comment:
This code block needs some big time comments. It is super icky :-/

Reply:
I've added comments, but maybe it would be better also to refactor it. Do you have any ideas on a cleaner impl of equivalent logic?

Reply:
Let me try to make time to understand exactly what is happening here now that you've added the comments.  Try to reply today.